File: LPIS_1/src/main/java/com/example/lab/expression/FuncCallExpression.java
package com.example.lab.expression;

import com.example.lab.*;
import com.example.lab.interpreter.InterpreterContext;
import com.example.lab.statement.StatementContext;

import java.util.List;

public record FuncCallExpression(String name, List<? extends Expression> args, int line) implements Expression {

    public FuncCallExpression(String name, int line) {
        this(name, List.of(), line);
    }

    @Override
    public Type getType(FunctionContext context) {
        var s = signature(context);
        return context.getReturnType(s).map(x -> x.toType()).orElse(null);
    }

    @Override
    public void runVoidInterpreter(InterpreterContext context) {
        context.callFunctionVoid(name, args.stream().map(x -> x.runInterpreter(context)).toList());
    }

    public FunctionSignature signature(FunctionContext context) {
        var argTypes = args.stream().map(x -> x.getType(context)).toList();
        return new FunctionSignature(name, argTypes);
    }

    @Override
    public boolean checkContextSemantic(StatementContext context) {
        if (!args().stream().map(x -> x.checkContextSemantic(context)).allMatch(x -> x))
            return false;
        var s = signature(context);
        if (!context.hasFunction(s))
            return SemanticError.print("call not define function signature " + s, line);
        var type = context.getReturnType(s).get();
        if (type == ReturnType.VOID)
            return SemanticError.print("get return on void function " + s, line);
        return Expression.super.checkContextSemantic(context);
    }

}

File: LPIS_1/src/main/java/com/example/lab/expression/FuncCallExpression.kt
package com.example.lab.expression

import com.example.lab.FunctionSignature
import com.example.lab.expression.provider.ExpressionProvider
import com.example.lab.interpreter.InterpreterContext
import com.example.lab.statement.StatementContext

data class FuncCallExpressionProvider(
	val name: String,
	val arguments: List<ExpressionProvider>
) : ExpressionProvider {

	override fun provide(context: StatementContext): Expression {
		return FuncCallExpression(name, arguments.map { it.provide(context) })
	}
}

data class FuncCallExpression(
	val name: String,
	val arguments: List<Expression>,
) : UnitExpression {

	private fun signature(context: ExpressionProvider.Context): FunctionSignature {
		val argTypes = arguments.map { it.getType(context) }
		return FunctionSignature(name, argTypes)
	}

	override fun getTypeOrNull(context: ExpressionProvider.Context) = context.getReturnType(signature(context))

	override fun runUnit(context: InterpreterContext) {
		context.callFunctionVoid(name, arguments.map { it.runAny(context) })
	}
}

private fun Expression.runAny(context: InterpreterContext) = when(this) {
	is ElementExpression -> runElement(context)
	is ListExpression -> runList(context)
	is NothingExpression -> runInterpreter(context)
	is UnitExpression -> runUnit(context)
	else -> TODO()
}